### ISSUE TITLE
automation of linting test using pre-commit

### ISSUE THREAD

**Issue by 12rambau** [sphinx-doc/sphinx#11205]
I was working on a PR and I was surprised to see many linting errors.
I didn't see any documentation on which test I should run from my side (ruff, flak8 isort others). 

**Describe the solution you'd like**
Use pre-commits to run all these linting checks automatically and use the same pre-commits in the github tests for consistency

**Comment by hugovk** [comment]
Another benefit is pinning the lint tools, so you don't get hit by sudden failures when linters release with add new checks.

**Comment by onerandomusername** [comment]
> Another benefit is pinning the lint tools, so you don't get hit by sudden failures when linters release with add new checks.

Yeah, I noticed that ruff is unpinned and will install whatever the latest version is. That's going to cause a lot of errors with how often ruff has a release with new features.

There's over 100 lint failures on master right now.

**Comment by 12rambau** [comment]
I started to work on it with the combo: black/prettier/ruff and it works quite ok.
Problem is taht lots of tests are now failing because of the linting. That make the test non robust to lint which is (according to me) an issue, I'll see if adding "pytest-regressions" and "beatifulsoup" could save us

**Comment by jdillard** [comment]
> Then we can update the versions and fix new warnings in a controlled manner in followups.

Is there a "best practice" for doing this? I wasn't sure if it is just setting a reminder or maybe a cron job that runs the latest version so you can monitor the number of new warnings and fix them before they get out of hand.

**Comment by 12rambau** [comment]
You can but do you really want to ? `ruff` and `black` are sending reminder in the pre-commit commands but to be fair once you set up something that works and more importantly that is consistent, you don't want to change it every month (unless you identify an error or want to add a new feature)

**Comment by hugovk** [comment]
I use https://pre-commit.ci - it autofixes PRs, and sends autoupdate PRs every month (by default, but I set it to quarterly with this at the end of `.pre-commit-config.yaml`:

```yaml
ci:
  autoupdate_schedule: quarterly
```

Also doesn't stop me running `pre-commit autoupdate`, which I might do if I'm also updating other stuff in the config file.

**Comment by 12rambau** [comment]
I'll go step by step as all the linting operation will crash. step one black + pre-commit + github actions

**Comment by hugovk** [comment]
This project uses tox, which is very well suited for adding an optional `tox -e lint` to run pre-commit.

For example:

https://github.com/termcolor/termcolor/blob/fe50c9a8e55b0fa8c5fc7d6503d749679cdf9a58/tox.ini#L24-L31

**Comment by AA-Turner** [comment]
Is this covered by these lines?

https://github.com/sphinx-doc/sphinx/blob/2a19d293b6ebd5a28ea4f6c148c4370ed8c01dee/tox.ini#L31-L41

I don't have `tox` set up so I can't easily test myself.

A

**Comment by hugovk** [comment]
Only partly, as a "standard" command that can be run.

But the benefit of pre-commit means those of us who like it can have instal the Git hook and have it automatically run (and autofix) when we run `git commit`.
